File: js/src/transport.ts
/**
 * Transport layer for Docker Sandboxes (`sbx`).
 *
 * {@link SbxTransport} is the seam between {@link SbxSandbox} and whatever
 * actually talks to Docker Sandboxes. {@link CliSbxTransport} shells out to the
 * `sbx` CLI, the only supported interface for local sandboxes. A cloud
 * transport (own REST client or the official `@docker/sandboxes` SDK) can be
 * dropped in later without touching `SbxSandbox`.
 *
 * Host commands are always spawned with an argv array (never a host shell), and
 * output is streamed with a hard cap: the child is killed the moment the cap is
 * reached rather than buffered and trimmed.
 */

import { spawn } from "node:child_process";
import {
  SbxAuthError,
  SbxCommandError,
  SbxError,
  SbxNotFoundError,
  SbxNotInstalledError,
  SbxPolicyError,
  SbxShapeError,
  SbxTimeoutError,
} from "./errors.js";

export const DEFAULT_MAX_OUTPUT_BYTES = 512_000;
/** Exit status reported by coreutils `timeout(1)` when it kills a command. */
const TIMEOUT_EXIT_CODE = 124;

/** Billable Docker Cloud Sandboxes shapes, in MiB. */
export const CLOUD_SHAPES: Record<string, { cpus: number; memoryMib: number }> = {
  micro: { cpus: 1, memoryMib: 2048 },
  small: { cpus: 2, memoryMib: 4096 },
  medium: { cpus: 4, memoryMib: 8192 },
  large: { cpus: 8, memoryMib: 16384 },
  xl: { cpus: 16, memoryMib: 32768 },
};

const CLOUD_DEFAULT_CPUS = 2;
const CLOUD_DEFAULT_MEMORY = "4g";

const MEMORY_UNITS: Record<string, number> = {
  b: 1,
  k: 1024,
  kb: 1024,
  ki: 1024,
  kib: 1024,
  m: 1024 ** 2,
  mb: 1024 ** 2,
  mi: 1024 ** 2,
  mib: 1024 ** 2,
  g: 1024 ** 3,
  gb: 1024 ** 3,
  gi: 1024 ** 3,
  gib: 1024 ** 3,
  t: 1024 ** 4,
  tb: 1024 ** 4,
  ti: 1024 ** 4,
  tib: 1024 ** 4,
};

/** Parse a binary memory string (`"4g"`, `"8192MiB"`, `"2048"`) to MiB. */
export function parseMemoryMib(memory: string): number | null {
  const text = memory.trim().toLowerCase();
  if (!text) return null;
  let index = text.length;
  while (index > 0 && /[a-z]/.test(text[index - 1] ?? "")) index -= 1;
  const number = text.slice(0, index);
  const unit = text.slice(index);
  const value = Number(number);
  if (!Number.isFinite(value)) return null;
  if (!unit) return Math.trunc(value); // bare numbers are MiB
  const multiplier = MEMORY_UNITS[unit];
  if (multiplier === undefined) return null;
  return Math.trunc((value * multiplier) / 1024 ** 2);
}

/** Map a `(cpus, memory)` pair onto a billable cloud shape name. */
export function resolveCloudShape(cpus?: number, memory?: string): string {
  const effectiveCpus = cpus ?? CLOUD_DEFAULT_CPUS;
  const effectiveMemory = memory ?? CLOUD_DEFAULT_MEMORY;
  const memoryMib = parseMemoryMib(effectiveMemory);
  if (memoryMib === null) throw new SbxShapeError(`Could not parse cloud memory value '${memory}'.`);
  for (const [name, shape] of Object.entries(CLOUD_SHAPES)) {
    if (shape.cpus === effectiveCpus && shape.memoryMib === memoryMib) return name;
  }
  const options = Object.entries(CLOUD_SHAPES)
    .map(([name, shape]) => `${name} (${shape.cpus} vCPU/${shape.memoryMib} MiB)`)
    .join(", ");
  throw new SbxShapeError(
    `${effectiveCpus} vCPU / ${effectiveMemory} is not a billable cloud shape. Valid shapes: ${options}.`,
  );
}

/** Raw result of one `sbx` invocation. */
export interface CommandResult {
  readonly argv: readonly string[];
  readonly output: string;
  readonly exitCode: number | null;
  readonly truncated: boolean;
}

/** One row of `sbx ls --json`. */
export interface SandboxInfo {
  readonly id: string;
  readonly name: string;
  readonly status?: string;
  readonly agent?: string;
  readonly raw: Record<string, unknown>;
}

export interface ExecOptions {
  timeout?: number;
  maxOutputBytes?: number;
}

export interface CreateOptions {
  agent?: string;
  workspace?: string;
  cpus?: number;
  memory?: string;
  profile?: string;
  pull?: string;
  /** Cloud-only time-to-live, e.g. `"2h"`. */
  ttl?: string;
  /** Cloud-only behaviour when `ttl` lapses: `"delete"` or `"stop"`. */
  onTimeout?: string;
}

export interface RemoveOptions {
  force?: boolean;
}

/** Abstract operations {@link SbxSandbox} needs from Docker Sandboxes. */
export interface SbxTransport {
  /** Whether this transport targets Docker Cloud Sandboxes. */
  readonly cloud?: boolean;
  exec(sandbox: string, command: string, options?: ExecOptions): Promise<CommandResult>;
  upload(sandbox: string, localPath: string, remotePath: string): Promise<CommandResult>;
  download(sandbox: string, remotePath: string, localPath: string): Promise<CommandResult>;
  create(name: string, options?: CreateOptions): Promise<CommandResult>;
  remove(name: string, options?: RemoveOptions): Promise<CommandResult>;
  list(): Promise<SandboxInfo[]>;
  inspect(name: string): Promise<Record<string, unknown> | null>;
  exists(name: string): Promise<boolean>;
  /** Cloud-only: current TTL/expiration. */
  ttl(sandbox: string): Promise<Record<string, unknown> | null>;
  /** Cloud-only: extend the TTL by `duration` (e.g. `"2h"`). */
  extendTtl(sandbox: string, duration: string): Promise<Record<string, unknown> | null>;
}

const AUTH_MARKERS = [
  "not logged in",
  "not authenticated",
  "you are not logged in",
  "please log in",
  "please run sbx login",
  "run 'sbx login'",
  'run "sbx login"',
  "authentication required",
  "unauthorized",
  "no credentials",
];

const POLICY_MARKERS = ["global network policy has not been initialized", "sbx policy init"];

const NOT_FOUND_MARKERS = ["no such sandbox", "sandbox not found", "no sandbox named", "does not exist", "not found"];

const LOGIN_HINT = "Run 'sbx login' first.";
const POLICY_HINT = "Run 'sbx policy init <allow-all|balanced|deny-all>' first.";

/** Map a non-zero {@link CommandResult} onto the most specific error. */
export function classifyFailure(result: CommandResult): SbxError {
  const text = result.output.toLowerCase();
  if (POLICY_MARKERS.some((marker) => text.includes(marker))) {
    return new SbxPolicyError(`${result.output.trim()}\n${POLICY_HINT}`);
  }
  if (AUTH_MARKERS.some((marker) => text.includes(marker))) {
    return new SbxAuthError(`${result.output.trim()}\n${LOGIN_HINT}`);
  }
  if (NOT_FOUND_MARKERS.some((marker) => text.includes(marker))) {
    return new SbxNotFoundError(result.output.trim() || "Sandbox not found.");
  }
  return new SbxCommandError(result.output.trim() || `Command failed with exit code ${result.exitCode}`, {
    argv: result.argv,
    exitCode: result.exitCode,
    output: result.output,
  });
}

/** Parse the JSON emitted by `sbx ls --json`. */
export function parseSandboxList(payload: string): SandboxInfo[] {
  const text = payload.trim();
  if (!text) return [];
  let data: unknown;
  try {
    data = JSON.parse(text);
  } catch (error) {
    throw new SbxCommandError(`Could not parse 'sbx ls --json' output: ${String(error)}`, { output: payload });
  }

  const records =
    data !== null && typeof data === "object" && !Array.isArray(data)
      ? ((data as Record<string, unknown>).sandboxes ?? [])
      : data;
  if (!Array.isArray(records)) return [];

  const infos: SandboxInfo[] = [];
  for (const record of records) {
    if (record === null || typeof record !== "object") continue;
    const raw = record as Record<string, unknown>;
    const name = String(raw.name ?? raw.id ?? "");
    if (!name) continue;
    infos.push({
      id: String(raw.id ?? name),
      name,
      status: raw.status !== undefined ? String(raw.status) : raw.state !== undefined ? String(raw.state) : undefined,
      agent: raw.agent !== undefined ? String(raw.agent) : undefined,
      raw,
    });
  }
  return infos;
}

interface RunOptions {
  timeout?: number;
  maxOutputBytes?: number;
}

export interface CliSbxTransportOptions {
  /** Extra environment variables merged over `process.env`. */
  env?: NodeJS.ProcessEnv;
  /**
   * When `true` (default), a sandbox-side `timeout` wraps long commands so the
   * remote process dies too. The host kill is always a backstop.
   */
  remoteTimeout?: boolean;
  /** Seconds the sandbox-side `timeout` waits after SIGTERM before SIGKILL. */
  remoteKillAfter?: number;
  /** Target Docker Cloud Sandboxes (`sbx --cloud …`) instead of local `sandboxd`. */
  cloud?: boolean;
}

/**
 * Transport that shells out to the `sbx` CLI.
 *
 * Works for local and Docker Cloud Sandboxes: `cloud: true` injects the global
 * `--cloud` flag so every verb is dispatched to the Cloud API.
 */
export class CliSbxTransport implements SbxTransport {
  readonly cloud: boolean;

  constructor(
    readonly binary: string = "sbx",
    private readonly options: CliSbxTransportOptions = {},
  ) {
    this.cloud = options.cloud ?? false;
  }

  private get remoteTimeout(): boolean {
    return this.options.remoteTimeout ?? true;
  }

  private get remoteKillAfter(): number {
    return Math.max(1, this.options.remoteKillAfter ?? 5);
  }

  private globalFlags(): string[] {
    return this.cloud ? ["--cloud"] : [];
  }

  private run(argv: string[], options: RunOptions): Promise<CommandResult> {
    const maxOutputBytes = options.maxOutputBytes ?? DEFAULT_MAX_OUTPUT_BYTES;
    const timeout = options.timeout !== undefined && options.timeout > 0 ? options.timeout : undefined;
    const full = [...this.globalFlags(), ...argv];

    return new Promise<CommandResult>((resolve, reject) => {
      const child = spawn(this.binary, full, {
        stdio: ["ignore", "pipe", "pipe"],
        env: this.options.env ? { ...process.env, ...this.options.env } : process.env,
      });

      const chunks: Buffer[] = [];
      let size = 0;
      let truncated = false;
      let timedOut = false;
      let settled = false;

      const finish = (fn: () => void): void => {
        if (settled) return;
        settled = true;
        fn();
      };

      const kill = (): void => {
        if (child.exitCode !== null || child.signalCode !== null) return;
        child.kill("SIGKILL");
      };

      const collect = (chunk: Buffer): void => {
        if (truncated) return;
        const remaining = maxOutputBytes - size;
        if (remaining <= 0) {
          truncated = true;
          kill();
          return;
        }
        if (chunk.length > remaining) {
          chunks.push(chunk.subarray(0, remaining));
          size = maxOutputBytes;
          truncated = true;
          kill();
          return;
        }
        chunks.push(chunk);
        size += chunk.length;
      };

      child.stdout?.on("data", collect);
      child.stderr?.on("data", collect);

      const timer =
        timeout !== undefined
          ? setTimeout(() => {
              timedOut = true;
              kill();
            }, timeout * 1000)
          : undefined;

      child.on("error", (error: NodeJS.ErrnoException) => {
        if (timer !== undefined) clearTimeout(timer);
        finish(() => {
          if (error.code === "ENOENT") {
            reject(
              new SbxNotInstalledError(
                `'${this.binary}' was not found on PATH. Install Docker Sandboxes and make sure the 'sbx' CLI is available.`,
              ),
            );
            return;
          }
          reject(new SbxError(`Failed to start '${this.binary}': ${error.message}`, { cause: error }));
        });
      });

      child.on("close", (code) => {
        if (timer !== undefined) clearTimeout(timer);
        const output = Buffer.concat(chunks).toString("utf8");
        finish(() => {
          if (timedOut) {
            reject(
              new SbxTimeoutError(
                `Command timed out after ${timeout}s: ${[this.binary, ...full].slice(0, 4).join(" ")} ...`,
                {
                  argv: [this.binary, ...full],
                  exitCode: null,
                  output,
                },
              ),
            );
            return;
          }
          resolve({ argv: [this.binary, ...full], output, exitCode: code, truncated });
        });
      });
    });
  }

  private async check(result: CommandResult): Promise<CommandResult> {
    if (result.exitCode !== 0) throw classifyFailure(result);
    return result;
  }

  private timeoutPrefix(timeout: number | undefined): string[] {
    if (!this.remoteTimeout || timeout === undefined || timeout <= 0) return [];
    return ["timeout", "-k", `${this.remoteKillAfter}s`, `${Math.trunc(timeout)}s`];
  }

  async exec(sandbox: string, command: string, options: ExecOptions = {}): Promise<CommandResult> {
    const timeout = options.timeout !== undefined && options.timeout > 0 ? options.timeout : undefined;
    const prefix = this.timeoutPrefix(timeout);
    const hostTimeout = prefix.length > 0 && timeout !== undefined ? timeout + this.remoteKillAfter + 1 : timeout;

    let result: CommandResult;
    try {
      result = await this.run(["exec", sandbox, ...prefix, "sh", "-c", command], {
        timeout: hostTimeout,
        maxOutputBytes: options.maxOutputBytes,
      });
    } catch (error) {
      if (error instanceof SbxTimeoutError) {
        throw new SbxTimeoutError(`Command timed out after ${timeout}s inside sandbox '${sandbox}'.`, {
          argv: error.argv,
          exitCode: null,
          output: error.output,
        });
      }
      throw error;
    }

    if (result.exitCode === TIMEOUT_EXIT_CODE && prefix.length > 0) {
      throw new SbxTimeoutError(`Command timed out after ${timeout}s inside sandbox '${sandbox}'.`, {
        argv: result.argv,
        exitCode: TIMEOUT_EXIT_CODE,
        output: result.output,
      });
    }
    return result;
  }

  async upload(sandbox: string, localPath: string, remotePath: string): Promise<CommandResult> {
    return this.check(await this.run(["cp", localPath, `${sandbox}:${remotePath}`], {}));
  }

  async download(sandbox: string, remotePath: string, localPath: string): Promise<CommandResult> {
    return this.check(await this.run(["cp", `${sandbox}:${remotePath}`, localPath], {}));
  }

  async create(name: string, options: CreateOptions = {}): Promise<CommandResult> {
    if (this.cloud) return this.createCloud(name, options);
    const args = ["create", "--name", name];
    if (options.cpus) args.push("--cpus", String(options.cpus));
    if (options.memory) args.push("--memory", options.memory);
    if (options.profile) args.push("--profile", options.profile);
    if (options.pull) args.push("--pull", options.pull);
    args.push(options.agent ?? "shell");
    if (options.workspace) args.push(options.workspace);
    return this.check(await this.run(args, {}));
  }

  private async createCloud(name: string, options: CreateOptions): Promise<CommandResult> {
    if (options.workspace) {
      throw new SbxError(
        "Cloud sandboxes have no host workspace; omit 'workspace' (download results with downloadFiles()).",
      );
    }
    const shape = resolveCloudShape(options.cpus, options.memory);
    const spec = CLOUD_SHAPES[shape];
    if (spec === undefined) throw new SbxShapeError(`Unknown cloud shape '${shape}'.`);
    const args = ["create", "--name", name, "--cpus", String(spec.cpus), "--memory", `${spec.memoryMib}m`];
    if (options.ttl) args.push("--ttl", options.ttl);
    if (options.onTimeout) args.push("--on-timeout", options.onTimeout);
    args.push(options.agent ?? "shell");
    return this.check(await this.run(args, {}));
  }

  async remove(name: string, options: RemoveOptions = {}): Promise<CommandResult> {
    const args = ["rm"];
    if (options.force ?? true) args.push("--force");
    args.push(name);
    return this.check(await this.run(args, {}));
  }

  async list(): Promise<SandboxInfo[]> {
    const result = await this.check(await this.run(["ls", "--json"], {}));
    return parseSandboxList(result.output);
  }

  async inspect(name: string): Promise<Record<string, unknown> | null> {
    if (this.cloud) {
      throw new SbxError("'sbx inspect' is not supported in cloud mode; use list() for cloud sandbox metadata.");
    }
    const result = await this.run(["inspect", name, "--json"], {});
    if (result.exitCode !== 0) {
      const error = classifyFailure(result);
      if (error instanceof SbxNotFoundError) return null;
      throw error;
    }
    return parseJsonObject(result);
  }

  async exists(name: string): Promise<boolean> {
    const infos = await this.list();
    return infos.some((info) => info.name === name || info.id === name);
  }

  async ttl(sandbox: string): Promise<Record<string, unknown> | null> {
    if (!this.cloud) throw new SbxError("TTL inspection is cloud-only; local sandboxes are not TTL-managed.");
    return parseJsonObject(await this.check(await this.run(["ttl", sandbox, "--json"], {})));
  }

  async extendTtl(sandbox: string, duration: string): Promise<Record<string, unknown> | null> {
    if (!this.cloud) throw new SbxError("TTL extension is cloud-only; local sandboxes are not TTL-managed.");
    const value = duration.startsWith("+") ? duration : `+${duration}`;
    return parseJsonObject(await this.check(await this.run(["ttl", value, sandbox, "--json"], {})));
  }
}

function parseJsonObject(result: CommandResult): Record<string, unknown> | null {
  const payload = result.output.trim();
  if (!payload) return null;
  try {
    const data: unknown = JSON.parse(payload);
    return data !== null && typeof data === "object" && !Array.isArray(data) ? (data as Record<string, unknown>) : null;
  } catch (error) {
    throw new SbxCommandError(`Could not parse sbx JSON output: ${String(error)}`, {
      argv: result.argv,
      exitCode: result.exitCode,
      output: result.output,
    });
  }
}

File: python/src/deepagents_sbx/transport.py
"""Transport layer for Docker Sandboxes (``sbx``).

:class:`SbxTransport` is the seam between :class:`~deepagents_sbx.backend.SbxSandbox`
and whatever actually talks to Docker Sandboxes. The first implementation,
:class:`CliSbxTransport`, shells out to the ``sbx`` CLI, which is the only
*supported* local interface (there is no documented local REST API).

Two independent axes are meant to stay swappable behind this interface:

* **environment** -- local vs. cloud;
* **implementation** -- CLI subprocess vs. an HTTP client vs. the official
  ``@docker/sandboxes`` TypeScript SDK.

Cloud transport (a hand-rolled REST client, or the official SDK from JS) can be
dropped in later without touching ``SbxSandbox``.

Design notes
------------
* Commands are ALWAYS passed as an argv array with ``shell=False`` on the host.
  The user's command is wrapped in a single ``sh -c`` *inside* the sandbox, so
  no host-side interpolation happens.
* Output is streamed and the subprocess is killed the moment ``max_output_bytes``
  is reached. It is never fully buffered and then trimmed -- a noisy command
  would otherwise be able to exhaust host memory.
"""

from __future__ import annotations

import contextlib
import json
import os
import shutil
import signal
import subprocess
import threading
from abc import ABC, abstractmethod
from collections.abc import Mapping, Sequence
from dataclasses import dataclass, field
from typing import Any

from .errors import (
    SbxAuthError,
    SbxCommandError,
    SbxError,
    SbxNotFoundError,
    SbxNotInstalledError,
    SbxPolicyError,
    SbxTimeoutError,
)

DEFAULT_MAX_OUTPUT_BYTES: int = 512_000
"""Default cap on captured command output, mirroring the Python backend's cap."""

_READ_CHUNK: int = 64 * 1024
_KILL_GRACE_SECONDS: float = 5.0
_BINARY: str = "sbx"
_TIMEOUT_EXIT_CODE: int = 124
"""Exit status reported by coreutils ``timeout(1)`` when it kills a command."""

_AUTH_MARKERS: tuple[str, ...] = (
    "not logged in",
    "not authenticated",
    "you are not logged in",
    "please log in",
    "please run sbx login",
    "run 'sbx login'",
    'run "sbx login"',
    "authentication required",
    "unauthorized",
    "no credentials",
)

_POLICY_MARKERS: tuple[str, ...] = (
    "global network policy has not been initialized",
    "sbx policy init",
)

_NOT_FOUND_MARKERS: tuple[str, ...] = (
    "no such sandbox",
    "sandbox not found",
    "no sandbox named",
    "does not exist",
    "not found",
)

_LOGIN_HINT = "Run 'sbx login' first."
_POLICY_HINT = "Run 'sbx policy init <allow-all|balanced|deny-all>' first."


@dataclass(frozen=True, slots=True)
class CommandResult:
    """Raw result of one ``sbx`` invocation."""

    argv: tuple[str, ...]
    output: str
    exit_code: int | None
    truncated: bool = False

    @property
    def ok(self) -> bool:
        """Whether the command exited zero."""
        return self.exit_code == 0


@dataclass(frozen=True, slots=True)
class SandboxInfo:
    """One row of ``sbx ls --json``.

    ``id`` is the stable per-sandbox identifier; ``name`` is the mutable display
    name. ``raw`` keeps the untouched record for callers that need a field this
    dataclass does not model yet.
    """

    id: str
    name: str
    status: str | None = None
    agent: str | None = None
    raw: Mapping[str, Any] = field(default_factory=dict)


def _ensure_binary(binary: str) -> None:
    """Fail fast when ``binary`` cannot be resolved.

    An absolute/relative path is trusted as-is; a bare name is looked up on
    ``PATH``. This is what turns a missing ``sbx`` into
    :class:`SbxNotInstalledError` instead of a bare ``FileNotFoundError``.
    """
    if os.sep in binary or (os.altsep and os.altsep in binary):
        return
    if shutil.which(binary) is None:
        raise SbxNotInstalledError(
            f"{binary!r} was not found on PATH. Install Docker Sandboxes and "
            "make sure the 'sbx' CLI is available (https://docs.docker.com/ai/sandboxes/)."
        )


def _kill(proc: subprocess.Popen[bytes]) -> None:
    """Best-effort SIGKILL of the subprocess and its process group."""
    if proc.poll() is not None:
        return
    with contextlib.suppress(Exception):
        if os.name == "posix":
            os.killpg(os.getpgid(proc.pid), signal.SIGKILL)
        else:  # pragma: no cover - exercised on Windows only
            proc.kill()
    with contextlib.suppress(Exception):
        proc.kill()


def _run(
    argv: Sequence[str],
    *,
    timeout: float | None = None,
    max_output_bytes: int = DEFAULT_MAX_OUTPUT_BYTES,
    env: Mapping[str, str] | None = None,
) -> CommandResult:
    """Run ``argv`` streaming its combined output, with a hard size cap.

    Args:
        argv: Argument array. Never a shell string.
        timeout: Host-side deadline in seconds; the process group is killed when
            it elapses.
        max_output_bytes: Once this many bytes have been captured the process is
            killed and ``truncated`` is set on the result.
        env: Optional environment override.

    Returns:
        A :class:`CommandResult` with decoded, merged stdout+stderr.

    Raises:
        SbxNotInstalledError: The binary is missing.
        SbxTimeoutError: The deadline elapsed.
    """
    argv = [str(arg) for arg in argv]
    _ensure_binary(argv[0])

    try:
        proc = subprocess.Popen(
            argv,
            stdin=subprocess.DEVNULL,
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            shell=False,
            env=None if env is None else {**os.environ, **env},
            start_new_session=(os.name == "posix"),
        )
    except FileNotFoundError as exc:
        raise SbxNotInstalledError(f"{argv[0]!r} could not be executed.") from exc
    except OSError as exc:
        raise SbxError(f"Failed to start {argv[0]!r}: {exc}") from exc

    buffer = bytearray()
    state = {"truncated": False}

    def _reader() -> None:
        stream = proc.stdout
        if stream is None:  # pragma: no cover - stdout is always a pipe here
            return
        with contextlib.suppress(OSError, ValueError):
            while True:
                chunk = stream.read(_READ_CHUNK)
                if not chunk:
                    return
                remaining = max_output_bytes - len(buffer)
                if remaining <= 0:
                    state["truncated"] = True
                    _kill(proc)
                    return
                if len(chunk) > remaining:
                    buffer.extend(chunk[:remaining])
                    state["truncated"] = True
                    _kill(proc)
                    return
                buffer.extend(chunk)

    reader = threading.Thread(target=_reader, name="sbx-reader", daemon=True)
    reader.start()

    timed_out = False
    exit_code: int | None
    try:
        exit_code = proc.wait(timeout=timeout)
    except subprocess.TimeoutExpired:
        timed_out = True
        _kill(proc)
        with contextlib.suppress(subprocess.TimeoutExpired):
            exit_code = proc.wait(timeout=_KILL_GRACE_SECONDS)
        if proc.poll() is None:  # pragma: no cover - SIGKILL is immediate
            exit_code = None
    finally:
        with contextlib.suppress(Exception):
            if proc.stdout is not None:
                proc.stdout.close()
        reader.join(timeout=_KILL_GRACE_SECONDS)

    output = bytes(buffer).decode("utf-8", errors="replace")
    if timed_out:
        raise SbxTimeoutError(
            f"Command timed out after {timeout}s: {' '.join(argv[:3])} ...",
            argv=tuple(argv),
            exit_code=None,
            output=output,
        )
    return CommandResult(
        argv=tuple(argv),
        output=output,
        exit_code=exit_code,
        truncated=state["truncated"],
    )


def classify_failure(result: CommandResult) -> SbxError:
    """Map a non-zero :class:`CommandResult` onto the most specific error.

    Detection is message-based because the ``sbx`` CLI does not emit stable
    machine-readable error codes. Anything unrecognized becomes a
    :class:`SbxCommandError` carrying argv/exit/output.
    """
    text = result.output.lower()
    if any(marker in text for marker in _POLICY_MARKERS):
        return SbxPolicyError(f"{result.output.strip()}\n{_POLICY_HINT}")
    if any(marker in text for marker in _AUTH_MARKERS):
        return SbxAuthError(f"{result.output.strip()}\n{_LOGIN_HINT}")
    if any(marker in text for marker in _NOT_FOUND_MARKERS):
        return SbxNotFoundError(result.output.strip() or "Sandbox not found.")
    return SbxCommandError(
        result.output.strip() or f"Command failed with exit code {result.exit_code}",
        argv=result.argv,
        exit_code=result.exit_code,
        output=result.output,
    )


class SbxTransport(ABC):
    """Abstract operations :class:`SbxSandbox` needs from Docker Sandboxes."""

    @abstractmethod
    def exec(
        self,
        sandbox: str,
        command: str,
        *,
        timeout: float | None = None,
        max_output_bytes: int = DEFAULT_MAX_OUTPUT_BYTES,
    ) -> CommandResult:
        """Run ``command`` inside ``sandbox`` via ``sh -c``."""

    @abstractmethod
    def upload(self, sandbox: str, local_path: str, remote_path: str) -> CommandResult:
        """Copy a host file into the sandbox."""

    @abstractmethod
    def download(self, sandbox: str, remote_path: str, local_path: str) -> CommandResult:
        """Copy a file out of the sandbox onto the host."""

    @abstractmethod
    def create(
        self,
        name: str,
        *,
        agent: str = "shell",
        workspace: str | None = None,
        cpus: int | None = None,
        memory: str | None = None,
        profile: str | None = None,
        pull: str | None = None,
    ) -> CommandResult:
        """Create a new sandbox."""

    @abstractmethod
    def remove(self, name: str, *, force: bool = True) -> CommandResult:
        """Delete a sandbox and its resources."""

    @abstractmethod
    def list(self) -> list[SandboxInfo]:
        """Return all sandboxes known to ``sbx``."""

    @abstractmethod
    def inspect(self, name: str) -> Mapping[str, Any] | None:
        """Return ``sbx inspect --json`` for ``name``, or ``None`` if absent."""

    def exists(self, name: str) -> bool:
        """Whether a sandbox with this name or stable id exists."""
        return any(name in (info.name, info.id) for info in self.list())


class CliSbxTransport(SbxTransport):
    """Local transport that shells out to the ``sbx`` CLI.

    Args:
        binary: Path or name of the CLI (default ``"sbx"``).
        env: Extra environment variables merged over ``os.environ``.
        remote_timeout: When ``True`` (default), a sandbox-side ``timeout`` wraps
            long commands so the *remote* process dies too. The host-side kill is
            always active as a backstop. Set to ``False`` on images without
            coreutils ``timeout``.
    """

    def __init__(
        self,
        binary: str = _BINARY,
        *,
        env: Mapping[str, str] | None = None,
        remote_timeout: bool = True,
        remote_kill_after: float = 5.0,
    ) -> None:
        self.binary = binary
        self.env = dict(env or {})
        self.remote_timeout = remote_timeout
        self.remote_kill_after = remote_kill_after

    # -- internals ---------------------------------------------------------

    def _run(
        self,
        args: Sequence[str],
        *,
        timeout: float | None = None,
        max_output_bytes: int = DEFAULT_MAX_OUTPUT_BYTES,
    ) -> CommandResult:
        return _run(
            [self.binary, *args],
            timeout=timeout,
            max_output_bytes=max_output_bytes,
            env=self.env or None,
        )

    def _check(self, result: CommandResult) -> CommandResult:
        if not result.ok:
            raise classify_failure(result)
        return result

    def _timeout_prefix(self, timeout: float | None) -> list[str]:
        """Argv prefix that wraps the command in the sandbox's coreutils ``timeout``.

        Returned as separate argv elements (``timeout -k 5s 120s``) rather than a
        nested shell string, so no additional quoting is needed: ``sbx exec``
        forwards each argv element verbatim to ``execve``.
        """
        if not self.remote_timeout or timeout is None or timeout <= 0:
            return []
        kill_after = max(1.0, float(self.remote_kill_after))
        return ["timeout", "-k", f"{kill_after:g}s", f"{int(timeout)}s"]

    # -- SbxTransport ------------------------------------------------------

    def exec(
        self,
        sandbox: str,
        command: str,
        *,
        timeout: float | None = None,
        max_output_bytes: int = DEFAULT_MAX_OUTPUT_BYTES,
    ) -> CommandResult:
        if timeout is not None and timeout <= 0:
            timeout = None
        prefix = self._timeout_prefix(timeout)
        # When the sandbox-side `timeout` wrapper is active it should fire first,
        # so give the host deadline a grace window; otherwise the host SIGKILL
        # would race the remote kill and could orphan the remote process.
        host_timeout = timeout
        if prefix and timeout is not None:
            host_timeout = timeout + max(1.0, float(self.remote_kill_after)) + 1.0
        try:
            result = self._run(
                ["exec", sandbox, *prefix, "sh", "-c", command],
                timeout=host_timeout,
                max_output_bytes=max_output_bytes,
            )
        except SbxTimeoutError as exc:
            # The host deadline carries the remote-kill grace; report the
            # caller's timeout, not the internal grace window.
            raise SbxTimeoutError(
                f"Command timed out after {timeout}s inside sandbox {sandbox!r}.",
                argv=exc.argv,
                exit_code=None,
                output=exc.output,
            ) from exc
        # When the sandbox-side wrapper is active, `timeout(1)` exits 124.
        if result.exit_code == _TIMEOUT_EXIT_CODE and prefix:
            raise SbxTimeoutError(
                f"Command timed out after {timeout}s inside sandbox {sandbox!r}.",
                argv=result.argv,
                exit_code=_TIMEOUT_EXIT_CODE,
                output=result.output,
            )
        return result

    def upload(self, sandbox: str, local_path: str, remote_path: str) -> CommandResult:
        return self._check(self._run(["cp", local_path, f"{sandbox}:{remote_path}"]))

    def download(self, sandbox: str, remote_path: str, local_path: str) -> CommandResult:
        return self._check(self._run(["cp", f"{sandbox}:{remote_path}", local_path]))

    def create(
        self,
        name: str,
        *,
        agent: str = "shell",
        workspace: str | None = None,
        cpus: int | None = None,
        memory: str | None = None,
        profile: str | None = None,
        pull: str | None = None,
    ) -> CommandResult:
        args: list[str] = ["create", "--name", name]
        if cpus:
            args += ["--cpus", str(int(cpus))]
        if memory:
            args += ["--memory", memory]
        if profile:
            args += ["--profile", profile]
        if pull:
            args += ["--pull", pull]
        args.append(agent)
        if workspace:
            args.append(workspace)
        return self._check(self._run(args))

    def remove(self, name: str, *, force: bool = True) -> CommandResult:
        args = ["rm"]
        if force:
            args.append("--force")
        args.append(name)
        return self._check(self._run(args))

    def list(self) -> list[SandboxInfo]:
        result = self._check(self._run(["ls", "--json"]))
        return parse_sandbox_list(result.output)

    def inspect(self, name: str) -> Mapping[str, Any] | None:
        result = self._run(["inspect", name, "--json"])
        if not result.ok:
            error = classify_failure(result)
            if isinstance(error, SbxNotFoundError):
                return None
            raise error
        try:
            data = json.loads(result.output)
        except json.JSONDecodeError as exc:
            raise SbxCommandError(
                f"Could not parse 'sbx inspect --json' output: {exc}",
                argv=result.argv,
                exit_code=result.exit_code,
                output=result.output,
            ) from exc
        return data if isinstance(data, Mapping) else None


def parse_sandbox_list(payload: str) -> list[SandboxInfo]:
    """Parse the JSON emitted by ``sbx ls --json``.

    Tolerates both a ``{"sandboxes": [...]}`` wrapper and a bare array, and both
    ``id`` and ``name`` fields being absent.
    """
    text = payload.strip()
    if not text:
        return []
    try:
        data = json.loads(text)
    except json.JSONDecodeError as exc:
        raise SbxCommandError(f"Could not parse 'sbx ls --json' output: {exc}", output=payload) from exc

    records: Any = data.get("sandboxes", []) if isinstance(data, Mapping) else data
    if not isinstance(records, Sequence):
        return []

    infos: list[SandboxInfo] = []
    for record in records:
        if not isinstance(record, Mapping):
            continue
        name = str(record.get("name") or record.get("id") or "")
        if not name:
            continue
        infos.append(
            SandboxInfo(
                id=str(record.get("id") or name),
                name=name,
                status=_optional_str(record.get("status") or record.get("state")),
                agent=_optional_str(record.get("agent")),
                raw=record,
            )
        )
    return infos


def _optional_str(value: Any) -> str | None:
    return None if value is None else str(value)


__all__ = [
    "DEFAULT_MAX_OUTPUT_BYTES",
    "CliSbxTransport",
    "CommandResult",
    "SandboxInfo",
    "SbxTransport",
    "classify_failure",
    "parse_sandbox_list",
]
